lms/djangoapps/shoppingcart/views.py

```diff
@@ -691,6 +691,16 @@ def postpay_callback(request):
         # See if this payment occurred as part of the verification flow process
         # If so, send the user back into the flow so they have the option
         # to continue with verification.
+
+        # Only orders where order_items.count() == 1 might be attempting to upgrade
+        attempting_upgrade = request.session.get('attempting_upgrade', False)
+        if attempting_upgrade and result['order'].has_items(CertificateItem):
+            course_id = result['order'].orderitem_set.all().select_subclasses()[0].course_id
```

Comment:
I still think this code isn't defensive enough.  It's true now that `attempting_upgrade` is set only when making a single-item purchase of a course product, but that won't necessarily be true if someone introduces a bug in the code.

I think this code should use `getattr()` to attempt to retrieve the course ID.  If the attribute isn't found, then skip firing the event.